File: 8vim/src/main/java/inc/flide/vim8/keyboardHelpers/KeyboardDataXmlParser.java
package inc.flide.vim8.keyboardHelpers;

import android.util.Xml;
import android.view.KeyEvent;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.io.InputStream;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import inc.flide.vim8.structures.Constants;
import inc.flide.vim8.structures.CustomKeycode;
import inc.flide.vim8.structures.FingerPosition;
import inc.flide.vim8.structures.KeyboardAction;
import inc.flide.vim8.structures.KeyboardActionType;
import inc.flide.vim8.structures.KeyboardData;
import inc.flide.vim8.structures.Quadrant;
import inc.flide.vim8.utils.QuadrantHelper;

class KeyboardDataXmlParser {
    private static final String KEYBOARD_DATA_TAG = "keyboardData";
    private static final String KEYBOARD_ACTION_TAG = "keyboardAction";
    private static final String KEYBOARD_ACTION_TYPE_TAG = "keyboardActionType";
    private static final String MOVEMENT_SEQUENCE_TAG = "movementSequence";
    private static final String INPUT_STRING_TAG = "inputString";
    private static final String INPUT_CAPSLOCK_STRING_TAG = "inputCapsLockString";
    private static final String INPUT_KEY_TAG = "inputKey";
    private static final String INPUT_KEY_FLAGS_TAG = "flags";
    private static final String INPUT_KEY_FLAG_TAG = "flag";
    private static final String LAYER_TAG = "layer";
    private static final String SECTOR_TAG = "sector";
    private static final String LAYER_LEVEL_ATTRIBUTE = "level";
    private static final String QUADRANT_ATTRIBUTE = "quadrant";
    private static final String POSITION_ATTRIBUTE = "position";

    private final XmlPullParser parser;

    KeyboardDataXmlParser(InputStream inputStream) throws XmlPullParserException, IOException {
        parser = Xml.newPullParser();
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
        parser.setInput(inputStream, null);
        parser.nextTag();
    }

    public KeyboardData readKeyboardData() throws IOException, XmlPullParserException {
        KeyboardData keyboardData = new KeyboardData();
        parser.require(XmlPullParser.START_TAG, null, KEYBOARD_DATA_TAG);
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }
            String tagName = parser.getName();
            switch (tagName) {
                case LAYER_TAG:
                    int layer = readLayerLevelAttribute();
                    readLayer(keyboardData, layer);
                    break;
            }
        }
        return keyboardData;
    }

    private void readLayer(KeyboardData keyboardData, int layer) throws IOException, XmlPullParserException {
        StringBuilder lowerCaseCharacters = new StringBuilder();
        StringBuilder upperCaseCharacters = new StringBuilder();

        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }
            String tagName = parser.getName();
            if (tagName.equals(SECTOR_TAG)) {
                Quadrant quadrant = readQuadrantAttribute();
                Objects.requireNonNull(quadrant);
                Map<List<FingerPosition>, KeyboardAction> actionMap =
                    readKeyboardActionMap(layer, quadrant, lowerCaseCharacters, upperCaseCharacters);
                keyboardData.addAllToActionMap(actionMap);
            }
        }
        if (layer >= Constants.DEFAULT_LAYER) {
            keyboardData.setLowerCaseCharacters(String.valueOf(lowerCaseCharacters), layer);
            keyboardData.setUpperCaseCharacters(String.valueOf(upperCaseCharacters), layer);
        }
    }

    private Map<List<FingerPosition>, KeyboardAction> readKeyboardActionMap(int layer, Quadrant quadrant, StringBuilder lowerCaseCharacters,
                                                                            StringBuilder upperCaseCharacters)
        throws IOException, XmlPullParserException {
        Map<List<FingerPosition>, KeyboardAction> keyboardActionMap = new HashMap<>();

        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }
            String tagName = parser.getName();
            if (tagName.equals(KEYBOARD_ACTION_TAG)) {
                int position = readPositionAttribute();
                int characterSetIndex = getCharacterSetIndex(quadrant, position);
                List<FingerPosition> movementSequence = QuadrantHelper.computeMovementSequence(layer, quadrant, position);
                Map.Entry<List<FingerPosition>, KeyboardAction> keyboardAction =
                    readKeyboardAction(layer, characterSetIndex, movementSequence, lowerCaseCharacters, upperCaseCharacters);
                keyboardActionMap.put(keyboardAction.getKey(), keyboardAction.getValue());
            }
        }
        return keyboardActionMap;
    }


    private Map.Entry<List<FingerPosition>, KeyboardAction> readKeyboardAction(int layer, int index, List<FingerPosition> computedMovementSequence,
                                                                               StringBuilder lowerCaseCharacters, StringBuilder upperCaseCharacters)
        throws XmlPullParserException, IOException {
        List<FingerPosition> movementSequence = null;
        KeyboardAction keyboardAction;
        KeyboardActionType keyboardActionType = null;
        String associatedText = "";
        String associatedCapsLockText = "";
        int keyEventCode = 0;
        int flags = 0;

        if (!computedMovementSequence.isEmpty()) {
            movementSequence = computedMovementSequence;
        }

        while (parser.next() != XmlPullParser.END_TAG) {

            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }

            String tagName = parser.getName();
            switch (tagName) {
                case KEYBOARD_ACTION_TYPE_TAG:
                    keyboardActionType = readKeyboardActionType();
                    break;
                case MOVEMENT_SEQUENCE_TAG:
                    movementSequence = readMovementSequence();
                    break;
                case INPUT_STRING_TAG:
                    associatedText = readInputString();
                    if (lowerCaseCharacters.length() == 0) {
                        lowerCaseCharacters.setLength(Constants.CHARACTER_SET_SIZE);
                    }
                    lowerCaseCharacters.setCharAt(index, associatedText.charAt(0));
                    break;
                case INPUT_KEY_TAG:
                    keyEventCode = readInputKey();
                    break;
                case INPUT_CAPSLOCK_STRING_TAG:
                    associatedCapsLockText = readInputCapsLockString();
                    if (upperCaseCharacters.length() == 0) {
                        upperCaseCharacters.setLength(Constants.CHARACTER_SET_SIZE);
                    }
                    upperCaseCharacters.setCharAt(index, associatedCapsLockText.charAt(0));
                    break;
                case INPUT_KEY_FLAGS_TAG:
                    flags = readInputFlags();
                    break;
                default:
                    //Logger.w(this, "keyboard_actions xml has unknown tag : " + tagName);
            }
        }

        keyboardAction = new KeyboardAction(keyboardActionType, associatedText, associatedCapsLockText, keyEventCode, flags, layer);
        return new AbstractMap.SimpleEntry<>(movementSequence, keyboardAction);
    }

    public Quadrant readQuadrantAttribute() {
        String quadrant = parser.getAttributeValue(null, QUADRANT_ATTRIBUTE);

        if (quadrant == null) {
            return null;
        }

        return Quadrant.valueOf(quadrant);
    }

    private int readLayerLevelAttribute() {
        String id = parser.getAttributeValue(null, LAYER_LEVEL_ATTRIBUTE);

        if (id == null) {
            return Constants.DEFAULT_LAYER;
        }

        try {
            return Integer.parseInt(id);
        } catch (Exception e) {
            return Constants.DEFAULT_LAYER;
        }
    }

    private int readPositionAttribute() {
        String position = parser.getAttributeValue(null, POSITION_ATTRIBUTE);

        if (position == null) {
            return Constants.DEFAULT_POSITION;
        }

        try {
            return Integer.parseInt(position) - 1;
        } catch (Exception e) {
            return Constants.DEFAULT_POSITION;
        }
    }

    private int readInputFlags() throws IOException, XmlPullParserException {

        int flags = 0;

        while (parser.next() != XmlPullParser.END_TAG) {

            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }

            String tagName = parser.getName();
            switch (tagName) {
                case INPUT_KEY_FLAG_TAG:
                    flags = flags | readInputFlag();
                    break;
                default:
                    //Logger.w(this, "keyboard_actions xml has unknown tag : " + tagName);
            }
        }
        return flags;
    }

    private int readInputFlag() throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, null, INPUT_KEY_FLAG_TAG);
        String inputKeyString = readText();
        parser.require(XmlPullParser.END_TAG, null, INPUT_KEY_FLAG_TAG);

        return Integer.parseInt(inputKeyString);
    }

    private int readInputKey() throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, null, INPUT_KEY_TAG);
        String inputKeyString = readText();
        parser.require(XmlPullParser.END_TAG, null, INPUT_KEY_TAG);

        //Strictly the inputKey has to has to be a Keycode from the KeyEvent class
        //Or it needs to be one of the customKeyCodes
        int keyCode = KeyEvent.keyCodeFromString(inputKeyString);
        if (keyCode == KeyEvent.KEYCODE_UNKNOWN) {
            try {
                keyCode = CustomKeycode.valueOf(inputKeyString).getKeyCode();
            } catch (IllegalArgumentException error) {
                keyCode = KeyEvent.KEYCODE_UNKNOWN;
            }
        }

        return keyCode;
    }

    private String readInputString() throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, null, INPUT_STRING_TAG);
        String inputString = readText();
        parser.require(XmlPullParser.END_TAG, null, INPUT_STRING_TAG);

        return inputString;
    }

    private String readInputCapsLockString() throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, null, INPUT_CAPSLOCK_STRING_TAG);
        String inputString = readText();
        parser.require(XmlPullParser.END_TAG, null, INPUT_CAPSLOCK_STRING_TAG);

        return inputString;
    }

    private List<FingerPosition> readMovementSequence() throws IOException, XmlPullParserException {

        parser.require(XmlPullParser.START_TAG, null, MOVEMENT_SEQUENCE_TAG);
        String movementSequenceString = readText();
        parser.require(XmlPullParser.END_TAG, null, MOVEMENT_SEQUENCE_TAG);

        String[] movementSequenceList = movementSequenceString.split("\\s*;\\s*");
        List<FingerPosition> movementSequence = new ArrayList<>();
        for (String movement : movementSequenceList) {
            movementSequence.add(FingerPosition.valueOf(movement));
        }
        return movementSequence;
    }

    private KeyboardActionType readKeyboardActionType() throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, null, KEYBOARD_ACTION_TYPE_TAG);
        String keyboardActionTypeString = readText();
        parser.require(XmlPullParser.END_TAG, null, KEYBOARD_ACTION_TYPE_TAG);

        return KeyboardActionType.valueOf(keyboardActionTypeString);
    }

    private String readText() throws IOException, XmlPullParserException {
        String result = "";
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.getText();
            parser.nextTag();
        }
        return result;
    }

    private int getCharacterSetIndex(Quadrant quadrant, int position) {
        int base = quadrant.ordinal() / 2 * 8;
        int delta = quadrant.ordinal() % 2;
        return base + position * 2 + delta;
    }
}

File: 8vim/src/main/java/inc/flide/vim8/keyboardHelpers/KeyboardDataXmlParser.kt
package inc.flide.vim8.keyboardHelpers

import android.util.Xml
import android.view.KeyEvent
import inc.flide.vim8.structures.*
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.io.InputStream
import java.util.*

internal class KeyboardDataXmlParser(inputStream: InputStream) {
    private val parser: XmlPullParser = Xml.newPullParser()

    @Throws(IOException::class, XmlPullParserException::class)
    fun readKeyboardData(): KeyboardData {
        val keyboardData = KeyboardData()
        parser.require(XmlPullParser.START_TAG, null, KEYBOARD_DATA_TAG)
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            when (parser.name) {
                KEYBOARD_CHARACTER_SET_TAG -> readKeyboardCharacterSet(keyboardData)
                KEYBOARD_ACTION_MAP_TAG -> keyboardData.setActionMap(readKeyboardActionMap())
            }
        }
        return keyboardData
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readKeyboardActionMap(): MutableMap<MutableList<FingerPosition>?, KeyboardAction?> {
        val keyboardActionMap: MutableMap<MutableList<FingerPosition>?, KeyboardAction?> = HashMap()
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            val tagName = parser.name
            if (tagName == KEYBOARD_ACTION_TAG) {
                val keyboardAction = readKeyboardAction()
                keyboardActionMap[keyboardAction.key] = keyboardAction.value
            }
        }
        return keyboardActionMap
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun readKeyboardCharacterSet(keyboardData: KeyboardData) {
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            when (parser.name) {
                KEYBOARD_CHARACTER_SET_LOWERCASE_TAG -> {
                    parser.require(XmlPullParser.START_TAG, null, KEYBOARD_CHARACTER_SET_LOWERCASE_TAG)
                    val keyboardLowerCaseCharacterSet = readText()
                    keyboardData.setLowerCaseCharacters(keyboardLowerCaseCharacterSet)
                    parser.require(XmlPullParser.END_TAG, null, KEYBOARD_CHARACTER_SET_LOWERCASE_TAG)
                }
                KEYBOARD_CHARACTER_SET_UPPERCASE_TAG -> {
                    parser.require(XmlPullParser.START_TAG, null, KEYBOARD_CHARACTER_SET_UPPERCASE_TAG)
                    val keyboardUpperCaseCharacterSet = readText()
                    keyboardData.setUpperCaseCharacters(keyboardUpperCaseCharacterSet)
                    parser.require(XmlPullParser.END_TAG, null, KEYBOARD_CHARACTER_SET_UPPERCASE_TAG)
                }
                else -> {}
            }
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun readKeyboardAction(): MutableMap.MutableEntry<MutableList<FingerPosition>?, KeyboardAction?> {
        var movementSequence: MutableList<FingerPosition>? = null
        val keyboardAction: KeyboardAction
        var keyboardActionType: KeyboardActionType? = null
        var associatedText: String? = ""
        var associatedCapsLockText: String? = ""
        var keyEventCode = 0
        var flags = 0
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            when (parser.name) {
                KEYBOARD_ACTION_TYPE_TAG -> keyboardActionType = readKeyboardActionType()
                MOVEMENT_SEQUENCE_TAG -> movementSequence = readMovementSequence()
                INPUT_STRING_TAG -> associatedText = readInputString()
                INPUT_KEY_TAG -> keyEventCode = readInputKey()
                INPUT_CAPSLOCK_STRING_TAG -> associatedCapsLockText = readInputCapsLockString()
                INPUT_KEY_FLAGS_TAG -> flags = readInputFlags()
                else -> {}
            }
        }
        keyboardAction = KeyboardAction(keyboardActionType, associatedText, associatedCapsLockText, keyEventCode, flags)
        return AbstractMap.SimpleEntry(movementSequence, keyboardAction)
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readInputFlags(): Int {
        var flags = 0
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            when (parser.name) {
                INPUT_KEY_FLAG_TAG -> flags = flags or readInputFlag()
                else -> {}
            }
        }
        return flags
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readInputFlag(): Int {
        parser.require(XmlPullParser.START_TAG, null, INPUT_KEY_FLAG_TAG)
        val inputKeyString = readText()
        parser.require(XmlPullParser.END_TAG, null, INPUT_KEY_FLAG_TAG)
        return Integer.valueOf(inputKeyString)
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readInputKey(): Int {
        parser.require(XmlPullParser.START_TAG, null, INPUT_KEY_TAG)
        val inputKeyString = readText()
        parser.require(XmlPullParser.END_TAG, null, INPUT_KEY_TAG)

        //Strictly the inputKey has to has to be a Keycode from the KeyEvent class
        //Or it needs to be one of the customKeyCodes
        var keyCode = KeyEvent.keyCodeFromString(inputKeyString)
        if (keyCode == KeyEvent.KEYCODE_UNKNOWN) {
            keyCode = try {
                CustomKeycode.valueOf(inputKeyString).getKeyCode()
            } catch (error: IllegalArgumentException) {
                KeyEvent.KEYCODE_UNKNOWN
            }
        }
        return keyCode
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readInputString(): String {
        parser.require(XmlPullParser.START_TAG, null, INPUT_STRING_TAG)
        val inputString = readText()
        parser.require(XmlPullParser.END_TAG, null, INPUT_STRING_TAG)
        return inputString
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readInputCapsLockString(): String {
        parser.require(XmlPullParser.START_TAG, null, INPUT_CAPSLOCK_STRING_TAG)
        val inputString = readText()
        parser.require(XmlPullParser.END_TAG, null, INPUT_CAPSLOCK_STRING_TAG)
        return inputString
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readMovementSequence(): MutableList<FingerPosition> {
        parser.require(XmlPullParser.START_TAG, null, MOVEMENT_SEQUENCE_TAG)
        val movementSequenceString = readText()
        parser.require(XmlPullParser.END_TAG, null, MOVEMENT_SEQUENCE_TAG)
        val movementSequenceList: Array<String> = movementSequenceString.split("\\s*;\\s*").toTypedArray()
        val movementSequence: MutableList<FingerPosition> = ArrayList()
        for (movement in movementSequenceList) {
            movementSequence.add(FingerPosition.valueOf(movement))
        }
        return movementSequence
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readKeyboardActionType(): KeyboardActionType {
        parser.require(XmlPullParser.START_TAG, null, KEYBOARD_ACTION_TYPE_TAG)
        val keyboardActionTypeString = readText()
        parser.require(XmlPullParser.END_TAG, null, KEYBOARD_ACTION_TYPE_TAG)
        return KeyboardActionType.valueOf(keyboardActionTypeString)
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readText(): String {
        var result = ""
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.text
            parser.nextTag()
        }
        return result
    }

    companion object {
        private const val KEYBOARD_DATA_TAG: String = "keyboardData"
        private const val KEYBOARD_ACTION_MAP_TAG: String = "keyboardActionMap"
        private const val KEYBOARD_ACTION_TAG: String = "keyboardAction"
        private const val KEYBOARD_ACTION_TYPE_TAG: String = "keyboardActionType"
        private const val MOVEMENT_SEQUENCE_TAG: String = "movementSequence"
        private const val INPUT_STRING_TAG: String = "inputString"
        private const val INPUT_CAPSLOCK_STRING_TAG: String = "inputCapsLockString"
        private const val INPUT_KEY_TAG: String = "inputKey"
        private const val INPUT_KEY_FLAGS_TAG: String = "flags"
        private const val INPUT_KEY_FLAG_TAG: String = "flag"
        private const val KEYBOARD_CHARACTER_SET_TAG: String = "keyboardCharacterSet"
        private const val KEYBOARD_CHARACTER_SET_LOWERCASE_TAG: String = "lowerCase"
        private const val KEYBOARD_CHARACTER_SET_UPPERCASE_TAG: String = "upperCase"
    }

    init {
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
        parser.setInput(inputStream, null)
        parser.nextTag()
    }
}